File: ios/Sources/HovrDevicePermissions/AppRuntimeCoordinator.swift
import UIKit

public final class AppRuntimeCoordinator {
    private weak var presenter: UIViewController?
    private let application: UIApplication
    private let options: RuntimeCoordinatorOptions
    private let alertPresenter: BlockingAlertPresenter
    private let networkCoordinator: NetworkConnectivityCoordinator?
    private let locationCoordinator: LocationPermissionCoordinator?
    private let notificationCoordinator: NotificationPermissionCoordinator?
    private var started = false

    public init(
        presenter: UIViewController,
        application: UIApplication,
        options: RuntimeCoordinatorOptions = RuntimeCoordinatorOptions()
    ) {
        self.presenter = presenter
        self.application = application
        self.options = options
        alertPresenter = BlockingAlertPresenter(presenter: presenter)
        networkCoordinator = options.monitorNetwork
            ? NetworkConnectivityCoordinator(presenter: presenter)
            : nil
        locationCoordinator = options.monitorLocation
            ? LocationPermissionCoordinator(
                presenter: presenter,
                alertPresenter: alertPresenter
            )
            : nil
        notificationCoordinator = options.monitorNotifications
            ? NotificationPermissionCoordinator(
                presenter: presenter,
                application: application,
                alertPresenter: alertPresenter
            )
            : nil
    }

    public func start() {
        guard !started else { return }
        started = true
        networkCoordinator?.attach()
        locationCoordinator?.attach()
        notificationCoordinator?.attach()
    }

    public func ensureAll() {
        networkCoordinator?.recheck()
        locationCoordinator?.ensureAccess()
        guard notificationCoordinator != nil else { return }
        Task {
            try? await Task.sleep(nanoseconds: PermissionLimits.notificationPromptDelayNanoseconds)
            await MainActor.run { [weak self] in
                self?.notificationCoordinator?.ensureAccess()
            }
        }
    }

    public func stop() {
        guard started else { return }
        started = false
        notificationCoordinator?.detach()
        locationCoordinator?.detach()
        networkCoordinator?.detach()
        alertPresenter.dismiss()
    }
}

File: android/hovr_device_permissions/src/main/kotlin/com/hovr/devicepermissions/AppRuntimeCoordinator.kt
package com.hovr.devicepermissions

import android.os.Handler
import android.os.Looper
import androidx.fragment.app.FragmentActivity
import com.hovr.devicepermissions.location.LocationPermissionCoordinator
import com.hovr.devicepermissions.network.NetworkConnectivityCoordinator
import com.hovr.devicepermissions.notification.NotificationPermissionCoordinator
import com.hovr.devicepermissions.ui.BlockingAlertPresenter

class AppRuntimeCoordinator(
    private val activity: FragmentActivity,
    private val options: RuntimeCoordinatorOptions = RuntimeCoordinatorOptions(),
) {
    private val mainHandler = Handler(Looper.getMainLooper())
    private val alertPresenter = BlockingAlertPresenter(activity)
    private val networkCoordinator =
        if (options.monitorNetwork) NetworkConnectivityCoordinator(activity) else null
    private val locationCoordinator =
        if (options.monitorLocation) {
            LocationPermissionCoordinator(activity, alertPresenter)
        } else {
            null
        }
    private val notificationCoordinator =
        if (options.monitorNotifications) {
            NotificationPermissionCoordinator(activity, alertPresenter)
        } else {
            null
        }
    private var attached = false

    fun attach() {
        if (attached) {
            return
        }
        attached = true
        networkCoordinator?.attach()
        locationCoordinator?.attach()
        notificationCoordinator?.attach()
    }

    fun detach() {
        if (!attached) {
            return
        }
        attached = false
        mainHandler.removeCallbacksAndMessages(null)
        notificationCoordinator?.detach()
        locationCoordinator?.detach()
        networkCoordinator?.detach()
        alertPresenter.dismiss()
    }

    fun ensureAll() {
        networkCoordinator?.recheck()
        locationCoordinator?.ensureAccess()
        if (notificationCoordinator != null) {
            mainHandler.postDelayed(
                { notificationCoordinator.ensureAccess() },
                PermissionLimits.NOTIFICATION_PROMPT_DELAY_MS,
            )
        }
    }
}
